File: src/classification/classifier.py
from collections import Counter

def classify_document(text, keyword_map):
    scores = Counter()
    matched_keywords = {}

    text_lower = text.lower()

    for category, keywords in keyword_map.items():
        for keyword in keywords:
            count = text_lower.count(keyword.lower())
            if count > 0:
                scores[category] += count
                matched_keywords.setdefault(category, {})[keyword] = count

    best_category = max(scores, key=scores.get) if scores and max(scores.values()) > 0 else "Uncategorized"

    return {
        "category": best_category,
        "scores": dict(scores),
        "matched_keywords": matched_keywords,
    }

File: php/src/Classifier.php
<?php

namespace Ibc;

/**
 * Port of src/classification/classifier.py
 */
class Classifier
{
    /**
     * @param array<string, string[]> $keywordMap
     * @return array{category: string, scores: array<string, int>, matched_keywords: array<string, array<string, int>>}
     */
    public static function classifyDocument(string $text, array $keywordMap): array
    {
        $scores = [];
        $matchedKeywords = [];
        $textLower = strtolower($text);

        foreach ($keywordMap as $category => $keywords) {
            foreach ($keywords as $keyword) {
                $count = substr_count($textLower, strtolower($keyword));
                if ($count > 0) {
                    $scores[$category] = ($scores[$category] ?? 0) + $count;
                    $matchedKeywords[$category][$keyword] = $count;
                }
            }
        }

        $bestCategory = 'Uncategorized';
        $bestCount = 0;
        foreach ($scores as $category => $count) {
            if ($count > $bestCount) {
                $bestCategory = $category;
                $bestCount = $count;
            }
        }

        return [
            'category' => $bestCategory,
            'scores' => $scores,
            'matched_keywords' => $matchedKeywords,
        ];
    }
}
